### ISSUE TITLE
offline mode: remote machine still need to download plugins from github

### ISSUE THREAD

**Issue by littenglt** [amitds1997/remote-nvim.nvim#83]
![image](https://github.com/amitds1997/remote-nvim.nvim/assets/146831861/7e84ff35-f117-4e7d-ad79-1981e872b836)
![image](https://github.com/amitds1997/remote-nvim.nvim/assets/146831861/ef8a29ed-5f51-4333-9b6c-79ab21e7c047)
![image](https://github.com/amitds1997/remote-nvim.nvim/assets/146831861/a18b4bf0-7ee1-48fb-987b-f6d152b99472)
that is unreasonable, right?
I set to offline mode because of the remote machine can't connect github normally

**Comment by amitds1997** [comment]
This is expected behavior (with the current implementation). This arises from your plugin manager (lazy.nvim) installing your plugins. Since there are a lot of plugin managers out there, there is no reliable way to ensure how to copy things over.

Maybe, I can add an extra option to specify people to specify directories in their `XDG_STATE_HOME` and `XDG_DATA_HOME` that they want to copy over to the remote as well. Then, we should be able to avoid this. 🤔

**Comment by littenglt** [comment]
after transfer nvim configurations, we can send ~/.local/share/nvim/ to remote machine directly?
what do you think?

**Comment by amitds1997** [comment]
> after transfer nvim configurations, we can send **~/.local/share/nvim/** to remote machine directly? what do you think?

Sounds fair. I would add a `*` to make sure that you can copy it over without having to specify individual directories ✅

**Comment by littenglt** [comment]
all the source code of plugins that needed locate on ~/.local/share/nvim

**Comment by amitds1997** [comment]
On my system, this is the state of `~/.local/share/nvim`
<img width="968" alt="image" src="https://github.com/amitds1997/remote-nvim.nvim/assets/29333147/655afe87-337c-4375-a237-34963b31c518">

All my plugins are stored by `lazy.nvim` under `lazy` folder.
`Mason` stores it's packages under `mason` but they are specifically for MacOS, so I'm not sure if they would be cross-os compatible.
There is `telescope_history` too which I can copy over to the remote but would not make sense for another machine because they are machine specific.

So, that's why I'm thinking configuration syntax of this type:
```lua
{
   data = "*", -- Copies everything inside the `~/.local/share/nvim`
   state = { "abc", "def" }, -- Copies only the specified directories over
}
```
One can use it to specify however one wishes to copy data over. So, this would support both yours and my use case.

Thanks for raising this issue btw, this would improve my quality of life too.

**Comment by littenglt** [comment]
it would be more elegent like below:
```
{
    scp_list = [
        '~/.config/nvim',
        ‘~/.local/sharenvim'
   ]
   exclude = []
}
```
if user select offline mode for remote machine
we just zip them seperately, scp, and unzip them to folder properly

**Comment by amitds1997** [comment]
Do you expect to copy over any other directories except your configuration and your plugins?

Why I kept it as 
```
{
   data = "*", -- Copies everything inside the `~/.local/share/nvim`
   state = { "abc", "def" }, -- Copies only the specified directories over
}
```
is because `data` corresponds to `vim.fn.stdpath("data")` and state corresponds to `vim.fn.stdpath("state")`. This helps me in determining exactly where to place the directories on the remote. This also helps in cleaning up the directories on the remote once you are done.

With `scp_list`, it gives me flexibility to copy over directories and specify things easily but it does not tell us where to place the copied over directories.

**Comment by amitds1997** [comment]
I have added feature to add this in the latest release (v0.2.3). Documentation for the same is available [here](https://github.com/amitds1997/remote-nvim.nvim?tab=readme-ov-file#copying-additional-directories-to-remote-neovim). If you still see any such issues, let me know.

For plugins, you would want to copy over your plugins directories onto your remote. Since you are using `lazy`, it might look something like this:
```lua
require('remote-nvim').setup({
   remote = {
     copy_dirs = {
       data = {
         dirs = { "lazy" },
         compression = {
           enabled = true,
         },
       },
     },
   },
})
```

In my tests, this was enough to not use internet at all and still make things work on the remote. This was done along with the offline-mode changes. I did observe that treesitter did not work unless I gave it internet access but that could be something specific to my configuration. Your mileage may vary. If you still see any issues, I'll be happy to look into it. Thanks!!
